src/esthri/src/ops/upload.rs

```diff
@@ -186,100 +193,149 @@ where
     upload_helper(s3, bucket, key, file, compressed).await
 }
 
-async fn create_file_chunk_stream<R>(
-    reader: R,
+// Creates a stream of smaller chunks for a larger chunk of data. This
+// is to reduce memory usage when uploading, as we currently do
+// multipart uploads of 8MB. Giving the rusuto uploading a stream that
+// it can consume causes less memory to be used than giving it whole 8MB
+// chunks as there is less data waiting around to be uploaded.
+fn create_stream_for_chunk<R>(
+    reader: Arc<Mutex<R>>,
     file_size: u64,
-) -> impl Stream<Item = Result<(i64, Vec<u8>)>>
+    chunk_number: u64,
+) -> (u64, impl Stream<Item = Result<Bytes>>)
 where
-    R: bio::Read + Send + 'static,
+    R: bio::Read + Send + bio::Seek + 'static,
 {
-    let upload_part_size = Config::global().upload_part_size();
-    let init_state = (file_size, /* part_number = */ 1, Some(reader));
-    Box::pin(stream::unfold(
+    let chunk_size = Config::global().upload_part_size();
+    let read_size = Config::global().upload_read_size();
+
+    // The total amount to read for this chunk
+    let to_read = u64::min(file_size - (chunk_size * chunk_number), chunk_size);
+
+    let init_state = (to_read, /* part_number = */ 0, Some(reader));
+
+    let stream = Box::pin(stream::unfold(
         init_state,
         move |(remaining, part_number, reader)| async move {
             let done_state = (0, 0, None);
             if remaining == 0 {
                 None
             } else {
-                let upload_part_size = u64::min(remaining, upload_part_size);
-                let mut buf = vec![0u8; upload_part_size as usize];
-                let result: Result<(Vec<u8>, usize, R)> = task::spawn_blocking(move || {
-                    let mut reader = reader.unwrap();
-                    let read_size = reader.read(&mut buf)?;
-                    Ok((buf, read_size, reader))
-                })
-                .await
-                .expect(EXPECT_SPAWN_BLOCKING);
+                let read_size = u64::min(remaining, read_size);
+
+                let mut buf = BytesMut::with_capacity(read_size as usize);
+                buf.resize(read_size as usize, 0);
+
+                let reader = reader.unwrap();
```

Comment:
Any danger of a panic here?

Reply:
No - this is heavily inspired/copied from your original code - `reader` is only `None` in the stop condition, so it'll always be `Some` here.